### ISSUE TITLE
Use liveness and readinessProbes

### ISSUE THREAD

**Issue by tilleryd** [DecipherNow/helm-charts#391]
So I admittedly don't know enough about `livenessProbes` and `readinessProbes`, but was just reading this article today (https://learnk8s.io/production-best-practices/) and it recommends not using `livenessProbes` if possible.  And recommends the two probes be different.

This is really more food for thought, but we should definitely start including readiness probes in all our containers, and making sure they fail properly so k8s can restart them.

_Originally posted by @chrisbsmith in https://github.com/DecipherNow/helm-charts/pull/380/files_

**Comment by danielpcox** [comment]
We actually ran into something during training on a few people's machines during the file-based service discovery part, where the liveness probes were failing and causing restarts of Control even though everything was OK.

**Comment by chrisbsmith** [comment]
That's interesting.  After looking at the liveliness probe, I would expect that if it failed there would be a real issue, or maybe there is a race condition when trying to do the liveliness check with a `tcpSocket` type.  Thanks for the info.

https://github.com/DecipherNow/helm-charts/blob/a38dcfa7a116616ae50f5f11081fa00b845a5802/control/templates/control.yaml#L41-L44

**Comment by whenhellfreezes** [comment]
Put up a PR for a catalog readiness probe. Doing down the list alphabetically atm any thought on the highest priority to add readiness probes to.

**Comment by kaitmore** [comment]
@chrisbsmith The livenessProbe is actually fine as is. It was failing during the workshop because we modified the control deployment to wait for 30s before starting up so that we could [copy in a file](https://github.com/DecipherNow/workshops/blob/master/training/4.%20Grey%20Matter%20Configuration/Grey%20Matter%20Configuration%20Training.md#flat-file):
```
command: ["/bin/sh", "-c", "sleep 30; /usr/local/bin/gm-control.sh"]
```

So, the livenessProbe was working as it should have. I just had people remove that part of the yaml during the workshop and it was no problem.

**Comment by kaitmore** [comment]
On another note,[ this is a good article ](https://blog.colinbreck.com/kubernetes-liveness-and-readiness-probes-how-to-avoid-shooting-yourself-in-the-foot/)on liveness/readiness probes and common pitfalls around using them.

**Comment by tilleryd** [comment]
From https://learnk8s.io/production-best-practices

> The Liveness probe is designed to restart your container when it's stuck.

Consider the following scenario: if your application is processing an infinite loop, there's no way to exit or ask for help.

When the process is consuming 100% CPU, it won't have time to reply to the (other) Readiness probe checks, and it will be eventually removed from the Service.

However, the Pod is still registered as an active replica for the current Deployment.

If you don't have a Liveness probe, it stays Running but detached from the Service.

In other words, not only is the process not serving any requests, but it is also consuming resources.

What should you do?

Expose an endpoint from your app
The endpoint always replies with a success response
Consume the endpoint from the Liveness probe
Please note that you should not use the Liveness probe to handle fatal errors in your app and request Kubernetes to restart the app.

Instead, you should let the app crash.

The Liveness probe should be used as a recovery mechanism only in case the process is not responsive.

>

**Comment by justincely** [comment]
These probes can be used for more than the services; they can also detect if the sidecar's successfully connected to the control plane: https://github.com/DecipherNow/gm-proxy/issues/489#issuecomment-569361939

See the full discussion there for an example and the full discussoin

**Comment by whenhellfreezes** [comment]
Going to add that probe to dashboard's proxy. That way if the globals are bad at least dashboard will catch it and so we can know what is up. Afterwards I think we should close this issue.
